feat(mount): Add mount info special inode

rolysr opened the pull request:
This commit introduces the initial implementation of a special inode to display information about the mountpoint. This feature will help in accessing client option values and initial parameters for external applications use.

rolysr committed:
feat(mount): Add mount info special inode

This commit introduces the initial implementation of a special inode
to display information about the mountpoint. This feature will help
in accessing client option values and initial parameters for external
applications use.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

rolysr committed:
refactor(mount): Separate mount info functions

This commit separates the mount info related functions into a new file
to improve the maintainability and organization of the code.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

rolysr committed:
feat(mount): Add tweaks update on mount info

Previous implementation of the mount info hidden file was not
showing the updated values corresponding to the client options that
could be changed by using the saunafs tweaks special file. This commit
allows to display the updated values on the mount info special inode.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

dmga44 commented on src/mount/fuse/main.cc:
```diff
@@ -175,6 +178,36 @@ int fuse_mnt_check_empty(const char *mnt, mode_t rootmode, off_t rootsize) {
+	std::tm *localtm = std::localtime(&now_time);
+
+	std::ostringstream oss;
+	oss << std::put_time(localtm, "%Y-%m-%d_%H:%M:%S");
```
If possible, consider moving the new functions, headers and other code changes in main.cc to the hidden file source code. The idea would be to be to leave here only the required changes.

dmga44 replied:
The only change which could be left here is a call to a function in sauna_client.h that stores the original command line.

rolysr replied:
Done

dmga44 commented on src/mount/fuse/main.cc:
```diff
@@ -266,6 +299,10 @@ static int mainloop(struct fuse_args *args, struct fuse_cmdline_opts *fuse_opts,
+
 	if (!gMountOptions.meta) {
 		SaunaClient::fs_init(params);
 	} else {
```
Consider building the parameters arguments for the hidden file in the call to fs_init().

rolysr replied:
Done

dmga44 commented on src/mount/mount_info.h:
```diff
@@ -0,0 +1,105 @@
+
+#include "mount/tweaks.h"
+
+typedef struct _mountInfo {
+	std::string startedDateUtc;
+#ifdef _WIN32
+	std::string sid;
+#else
+	std::string uid;
+	std::string gid;
+#endif
+	std::string username;
+	int pid;
+	std::string version;
+	std::string commitId;
+	std::string arguments;
+	std::unique_ptr<std::map<std::string, std::string>> mountOptions;
+} MountInfo;
```
Please modernize the _mountInfo to a class.

rolysr replied:
Done

dmga44 commented on src/mount/readdata.cc:
```diff
@@ -618,6 +619,7 @@ void read_data_init(uint32_t retries,
 	gPrefetchXorStripes = prefetchXorStripes;
 	gBandwidthOveruse = bandwidth_overuse;
 	gTweaks.registerVariable("PrefetchXorStripes", gPrefetchXorStripes);
+	tweakByMountOption["sfsprefetchxorstripes"] = "PrefetchXorStripes";
```
Please pass the original name of the variable as a param to the registerVariable function to simplify the implementation.

rolysr replied:
Done

dmga44 commented on src/mount/readdata.cc:
```diff
@@ -629,13 +631,21 @@ void read_data_init(uint32_t retries,
 	pthread_attr_destroy(&thattr);
 
 	gTweaks.registerVariable("ReadMaxRetries", maxRetries);
+	tweakByMountOption["sfsioretries"] = "ReadMaxRetries";
```
Please notice you're registering twice the "sfsioretries" option in tweakByMountOption.

For the specific case of "ReadMaxRetries" and "WriteMaxRetries", please consider doing the following:
- register them, but instead of "sfsioretries", as "sfsreadretries" and "sfswriteretries".
- modify the mountInfo.mountOptions got while building the mount info string to search for the last two mentioned keys (insert instead of "sfsioretries", the new "sfsreadretries" and "sfswriteretries").

rolysr replied:
Done

dmga44 reviewed (commented):
Great job, please see my comments.

rolysr committed:
refactor(mount): Refactor mount info functions

This change improves the separation of the mount info special
inode related functions, enhancing readability and maintainability.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>

rolysr committed:
feat(mount): Add thread safety to mount info

The previous implementation of the mount info hidden file did
not ensure thread-safe reading operations. This change introduces a
proper thread safety mechanism to protect the mount info special
inode during concurrent operations.

Signed-off-by: Rolando Sánchez Ramos <rolysr@leil.io>
